backend/s3/s3.go

```diff
@@ -4623,6 +4623,13 @@ func (f *Fs) Copy(ctx context.Context, src fs.Object, remote string) (fs.Object,
 		}
 		setFrom_s3CopyObjectInput_s3PutObjectInput(&req, ui.req)
 		req.MetadataDirective = aws.String(s3.MetadataDirectiveReplace)
+	} else if strings.EqualFold(srcPath, dstPath) { // prevent "trying to copy an object to itself..." error
```

Comment:
I'm not sure about this fix. S3 is supposed to be case insensitive so how does this manifest? Is it in `serve s3` on macOS maybe?

Also you are allowed to copy objects over themselves normally - that is how you set the metadata.

How do you reproduce the problem that this fixes?

Reply:
Here's an example of the issue from the integration tests: https://pub.rclone.org/integration-tests/2024-04-07-010011/s3-cmd.bisync-TestS3-5.txt (`ctrl+F` the word `InvalidRequest` to jump to the error)

So I think that rules out the possibility that it is specific to mac or `serve s3`.

I can reproduce it locally with AWS, and have verified that the same test passes with this patch. But that doesn't mean my patch is the correct fix 🙂 

It is possible that this issue affects only a subset of equal-folding names (like maybe unicode normalization form differences but not simple ASCII casing differences)

Reply:
AWS S3 [should ignore unicode normalization and case](https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html).

I managed to extract the request and response from the tests using

    go test -v -remote TestS3: -debug -verbose -run '^TestBisyncRemoteRemote$/^(extended_char_paths|normalization)$' -dump-bodies 2>&1 | tee /tmp/bisync-s3.log


```
2024/04/09 11:55:35 DEBUG : HTTP REQUEST (req 0xc000bb4400)
2024/04/09 11:55:35 DEBUG : PUT /bisync-115522-pacup/normalization/path2/folder/e%CC%81e%CC%81e%CC%81o%CC%88.txt HTTP/1.1
Host: rclone-test-vocekot3vekazev9picehox8.s3.eu-west-2.amazonaws.com
User-Agent: rclone/
Content-Length: 0
Authorization: XXXX
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Copy-Source: rclone-test-vocekot3vekazev9picehox8/bisync-115522-pacup/normalization/path2/folder/e%CC%81e%CC%81e%CC%81o%CC%88.txt
X-Amz-Date: 20240409T115535Z
X-Amz-Metadata-Directive: COPY
Accept-Encoding: gzip

2024/04/09 11:55:35 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2024/04/09 11:55:35 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2024/04/09 11:55:35 DEBUG : HTTP RESPONSE (req 0xc0004b2800)
2024/04/09 11:55:35 DEBUG : HTTP/1.1 400 Bad Request
Connection: close
Transfer-Encoding: chunked
Content-Type: application/xml
Date: Tue, 09 Apr 2024 11:55:34 GMT
Server: AmazonS3
X-Amz-Id-2: aPQJqzOJW77MdBbSPC4JR2gC4FlP8SPMs1WZa/vfdXH2b/d75KQOJp5QrgYteGfBFXpVbiWxeIY=
X-Amz-Request-Id: 83HASN30WZKYJPC5

179
<Error><Code>InvalidRequest</Code><Message>This copy request is illegal because it is trying to copy an object to itself without changing the object's metadata, storage class, website redirect location or encryption attributes.</Message><RequestId>83HASN30WZKYJPC5</RequestId><HostId>aPQJqzOJW77MdBbSPC4JR2gC4FlP8SPMs1WZa/vfdXH2b/d75KQOJp5QrgYteGfBFXpVbiWxeIY=</HostId></Error>
```

Here you can see that the object mentioned in the `PUT` and the object in the `X-Amz-Copy-Source` are actually the same, which is why AWS is complaining (here is me proving it to myself with cut and paste and python!)

```
>>> "e%CC%81e%CC%81e%CC%81o%CC%88.txt" == "e%CC%81e%CC%81e%CC%81o%CC%88.txt"
True
```

Hopefully that will give you something to dig into - I have to go for a lunch appointment but can help more this afternoon.

Reply:
I don't think changing this line is correct

https://github.com/rclone/rclone/blob/2ab2ec29f9230f2959ae1e4428b07a4c7080aaed/fs/operations/operations.go#L485

If we do that then Move will no longer obey its contract

https://github.com/rclone/rclone/blob/2ab2ec29f9230f2959ae1e4428b07a4c7080aaed/fs/operations/operations.go#L400-L401

Where is this called from? I suspect the caller should be setting `dst` to `nil`.

Reply:
It is called from [here](https://github.com/rclone/rclone/blob/04128f97ee2444253b636ed52c4138d1e754b8f3/fs/sync/sync.go#L389), and I don't know if passing `nil` there is the right fix either, due to this warning:

https://github.com/rclone/rclone/blob/04128f97ee2444253b636ed52c4138d1e754b8f3/fs/operations/operations.go#L403-L406

Reply:
What calls `operations.Move`? I made a stack trace and here is the answer

```
Stack trace:
goroutine 6818 [running]:
github.com/rclone/rclone/fs/log.Stack({0x206f080, 0xc000acc120}, {0xc000a6f900, 0x4a9})
	rclone/rclone/fs/log/log.go:86 +0x6e
github.com/rclone/rclone/fs/operations.move({0x2714728, 0xc000bab590}, {0x2728090?, 0xc000c9e2c0?}, {0x2728100, 0xc000acc120}, {0xc00124c1e1, 0x17}, {0x2728100, 0xc000acc120}, ...)
	rclone/rclone/fs/operations/operations.go:427 +0x1e8
github.com/rclone/rclone/fs/operations.Move(...)
	rclone/rclone/fs/operations/operations.go:414
github.com/rclone/rclone/fs/sync.(*syncCopyMove).pairChecker(0xc00070c500, 0x0?, 0x0?, 0xffffffffffffffff?, 0xc000bf5f98?)
	rclone/rclone/fs/sync/sync.go:389 +0x445
created by github.com/rclone/rclone/fs/sync.(*syncCopyMove).startCheckers in goroutine 5104
	rclone/rclone/fs/sync/sync.go:512 +0x45
```

I think the fundamental problem is that the `dst` object should point to an already existing **destination** but it is pointing to the **source** object I think.  That is `dst.Remote()` should be equal to `remote`.

The source of the call is here

https://github.com/rclone/rclone/blob/04128f97ee2444253b636ed52c4138d1e754b8f3/fs/sync/sync.go#L389

And you can see that it is quite clearly being called with the same object as `src` and `dst`. I think if we replace `dst` with `nil` here all will be well.

I made a fix for that here 8584bef00614a5642478b9ea4935d2ccd2492661

[v1.67.0-beta.7864.8584bef00.fix-7743-case-norm-sync](https://beta.rclone.org/branch/fix-7743-case-norm-sync/v1.67.0-beta.7864.8584bef00.fix-7743-case-norm-sync/) on branch [fix-7743-case-norm-sync](https://github.com/rclone/rclone/tree/fix-7743-case-norm-sync) (uploaded in 15-30 mins)

I'd appreciate your thoughts before I merge it though.

Reply:
I forgot to say, this fixes the bisync s3 tests above

    go test -v -remote TestS3:  -run '^TestBisyncRemoteRemote$/^(extended_char_paths|normalization)$'

Reply:
I think this works for `s3` because it's not case-insensitive, but this would break things for backends that actually are case-insensitive. For example, this fails for me:

```
go test ./fs/sync -v -remote TestDropbox:  -run '^TestFixCase$'
```

I think what makes this confusing is that on a case-insensitive backend, `hello` and `HELLO` actually _are_ considered the same remote, so if `hello` exists then the dst `HELLO` must also exist, even though it really doesn't. I think that's why the `src` and `dst` were passed as the same object -- because looking for the `dst` will end up finding the `src`. And so if we pass `nil` then we are doing what the comment warned about -- passing `nil` when the destination does exist, according to case-insensitive logic.

Reply:
With this additional change it seems to work: https://github.com/nielash/rclone/commit/cec22a5827dc172eb8b082f1336ac37388116386 

Maybe that's the missing piece?

Reply:
Yes I think that looks correct. Do you want to pull all the remaining bits into this PR and we can merge it?

Reply:
Will do.